File: app/src/main/java/com/wojdor/hearthstonecards/data/service/mapper/ZipResultMapper.java
package com.wojdor.hearthstonecards.data.service.mapper;

import com.wojdor.hearthstonecards.data.service.model.CardModel;

import java.util.ArrayList;
import java.util.List;

public class ZipResultMapper {

    private static final String HERO_CARD = "hero";

    private ZipResultMapper() {
    }

    public static List<CardModel> map(Object[] objects) {
        List<CardModel> allCards = new ArrayList<>();
        for (Object cardModels : objects) {
            if (isListOfCardModels(cardModels)) {
                ignoreHeroCards(allCards, (List<CardModel>) cardModels);
            }
        }
        return allCards;
    }

    private static void ignoreHeroCards(List<CardModel> allCards, List<CardModel> cardModels) {
        for (CardModel cardModel : cardModels) {
            if (isHeroCard(cardModel)) continue;
            allCards.add(cardModel);
        }
    }

    private static boolean isListOfCardModels(Object cardModels) {
        return cardModels instanceof List
                && !((List) cardModels).isEmpty()
                && ((List) cardModels).get(0) instanceof CardModel;
    }

    private static boolean isHeroCard(CardModel cardModel) {
        return cardModel.getCardId().toLowerCase().contains(HERO_CARD);
    }
}

File: app/src/main/java/com/wojdor/hearthstonecards/data/service/mapper/ZipResultMapper.kt
package com.wojdor.hearthstonecards.data.service.mapper

import com.wojdor.hearthstonecards.data.service.model.CardModel

object ZipResultMapper {

    private const val HERO_CARD = "hero"

    fun mapToCardModels(objects: Array<Any>): List<CardModel> {
        val allCards = mutableListOf<CardModel>()
        val cardModels = objects.filter { isListOfCardModels(it) }
        cardModels.forEach { ignoreHeroCards(allCards, it as List<CardModel>) }
        return allCards
    }

    private fun ignoreHeroCards(allCards: MutableList<CardModel>, cardModels: List<CardModel>) {
        allCards.addAll(cardModels.filterNot { isHeroCard(it) })
    }

    private fun isListOfCardModels(cardModels: Any): Boolean {
        return (cardModels is List<*>
                && cardModels.isNotEmpty()
                && cardModels.first() is CardModel)
    }

    private fun isHeroCard(cardModel: CardModel): Boolean {
        return cardModel.cardId.toLowerCase().contains(HERO_CARD)
    }
}
